In [ ]:
#importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as mpl
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier


In [ ]:
#reading and analysing credit card csv file
data=pd.read_csv('creditcard.csv')
data.head(10)
data.info()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5   2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6   4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7   7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8   7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9   9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   
5  0.260314 -0.568671  ... -0.208254 -0.559825 -0.026398 -0.371427 -0.232794   
6  0.081213  0.464960  ... -0.167716 -0.270710 -0.154104 -0.780055  0.750137   
7 -3.807864  0.615375  ...  1.943465 -1.015455  0.057504 -0.649709 -0.415267   
8  0.851084 -0.392048  ... -0.073425 -0.268092 -0.204233  1.011592  0.373205   
9  0.069539 -0.736727  ... -0.246914 -0.633753 -0.120794 -0.385050 -0.069733   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  
5  0.105915  0.253844  0.081080    3.67      0  
6 -0.257237  0.034507  0.005168    4.99      0  
7 -0.051634 -1.206921 -1.085339   40.80      0  
8 -0.384157  0.011747  0.142404   93.20      0  
9  0.094199  0.246219  0.083076    3.68      0  

[10 rows x 31 columns]

In [ ]:
#checking for null values
data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
#replacing null values with the mean of the column
data2=data.fillna(data.mean(numeric_only=True))
data2.isnull().sum()

In [ ]:
#converting the Class column into integer
data2['Class']=data2['Class'].astype(int)

In [ ]:
#dividing the dataset into train and test and scaling them
x=data2.drop("Class",axis='columns')
y=data2["Class"]
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)
ss=StandardScaler()
x_train=ss.fit_transform(x_train)
x_test=ss.fit_transform(x_test)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 30 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [ ]:
#implementing Random Forest Classifier
model_RFC=RandomForestClassifier(n_estimators=10,criterion='entropy')
model_RFC=model_RFC.fit(x_train,y_train)

In [ ]:
#predicting using Random Forest Classifier
y_pred_RFC=model_RFC.predict(x_test)

dtype('int64')

In [ ]:
#testing the accuracy of the Random Forest Classifier model
print("Accuracy:", accuracy_score(y_test, y_pred_RFC))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_RFC))
print("\nClassification Report:\n", classification_report(y_test, y_pred_RFC))

Accuracy: 0.9994382219725431

Confusion Matrix:
 [[56855     6]
 [   26    75]]

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     56861
           1       0.93      0.74      0.82       101

    accuracy                           1.00     56962
   macro avg       0.96      0.87      0.91     56962
weighted avg       1.00      1.00      1.00     56962



In [ ]:
# Implementing the Decision Tree Classifier model
model_tree=DecisionTreeClassifier(criterion='entropy')
model_tree=model_tree.fit(x_train,y_train)

In [ ]:
#Predicting using Decision Tree Classifier model
y_pred_tree=model_tree.predict(x_test)

In [ ]:
#Testing the accuracy of Decision Tree Classifier model
print("Accuracy:", accuracy_score(y_test, y_pred_tree))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_tree))
print("\nClassification Report:\n", classification_report(y_test, y_pred_tree))

Accuracy: 0.9991046662687406

Confusion Matrix:
 [[56832    29]
 [   22    79]]

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     56861
           1       0.73      0.78      0.76       101

    accuracy                           1.00     56962
   macro avg       0.87      0.89      0.88     56962
weighted avg       1.00      1.00      1.00     56962



In [ ]:
#Implementing KNeighborsClassifier model
model_k=KNeighborsClassifier(n_neighbors=5,metric='minkowski',p=2)
model_k.fit(x_train,y_train)

KNeighborsClassifier()

In [ ]:
#Predicting using KNeighborsClassifier model
y_pred_k=model_k.predict(x_test)

In [ ]:
#Testing accuracy of KNeighborsClassifier model
print("Accuracy:", accuracy_score(y_test, y_pred_k))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred_k))
print("\nClassification Report:\n", classification_report(y_test, y_pred_k))

Accuracy: 0.9994908886626171

Confusion Matrix:
 [[56853     8]
 [   21    80]]

Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00     56861
           1       0.91      0.79      0.85       101

    accuracy                           1.00     56962
   macro avg       0.95      0.90      0.92     56962
weighted avg       1.00      1.00      1.00     56962



In [ ]:
#Checking which model gives the best accuracy
macc=max(accuracy_score(y_test, y_pred_RFC),accuracy_score(y_test, y_pred_tree),accuracy_score(y_test, y_pred_k)) * 100
print(macc.round(4),"%")

99.9491 %


KNeighborsClassifier model gives the best accuracy with an accuracy of 99.9491%